akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala

```diff
@@ -344,11 +343,23 @@ private[client] object NewHostConnectionPool {
               slot.onConnectionFailed(ex)
             }
 
-          def onPull(): Unit = () // FIXME: do we need push / pull handling?
+          def onPull(): Unit = () // emitRequests makes sure not to push too early
 
           override def onDownstreamFinish(): Unit =
             withSlot(_.debug("Connection cancelled"))
 
+          /** Helper that makes sure requestOut is pulled before pushing */
+          private def emitRequest(request: HttpRequest): Unit =
+            if (requestOut.isAvailable) requestOut.push(request)
+            else
+              requestOut.setHandler(new OutHandler {
+                def onPull(): Unit = {
+                  requestOut.push(request)
+                  requestOut.setHandler(connection) // implicit assumption is that connection was the handler before
```

Comment:
Perhaps we should make this assumption explicit and fail fast by moving this inner `OutHandler` class to its own type and checking this existing handler is not of that type before doing `requestOut.setHandler(new ...)` above?

Reply:
There's no way to get the current handler for `SubSourceOutlet` (see https://github.com/akka/akka/issues/24158 for other inconsistencies between `SubSourceOutlet` and regular ones). I'll update the comment.